Race Condition in WebSocketConnection - Job failed: java.lang.IllegalStateException: FILLING_AND_PARSING

lachlan-roberts opened the pull request:
After PR https://github.com/jetty/jetty.project/pull/13264 there is a new race condition that can prevent a `WebSocketConnection` from completing.

```
2025-06-30 16:19:10.123:WARN :oejut.QueuedThreadPool:qtp-LocalServer-827: Job failed
java.lang.IllegalStateException: FILLING_AND_PARSING
	at org.eclipse.jetty.websocket.core.WebSocketConnection.fillAndParse(WebSocketConnection.java:489)
	at org.eclipse.jetty.websocket.core.WebSocketConnection.run(WebSocketConnection.java:389)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:981)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1211)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1166)
	at java.base/java.lang.Thread.run(Thread.java:840)
```

This stacktrace can be seen in `jetty-12.0.x` HEAD on build job for JDK17 and JDK21 (doesn't happen in JDK24) -
* https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/jetty-12.0.x/2565/pipeline/12

lachlan-roberts committed:
remove unnecessary locks around the WebSocketConnection networkBuffer

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

gregw reviewed (changes requested):
@lachlan-roberts This failure looks like more than a flake:
```
2025-06-29 20:48:43.421:WARN :oejuts.AdaptiveExecutionStrategy:WebSocket@2039328294-4500: Task run failed
java.lang.IllegalStateException: FILLING_AND_PARSING
	at org.eclipse.jetty.websocket.core.WebSocketConnection.fillAndParse(WebSocketConnection.java:480)
	at org.eclipse.jetty.websocket.core.WebSocketConnection.onFillable(WebSocketConnection.java:372)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:322)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:99)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:480)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:443)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:293)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.run(AdaptiveExecutionStrategy.java:201)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:311)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:981)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1211)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1166)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

lachlan-roberts committed:
PR #13299 - fix potential ISE from registering fillInterested in wrong state

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts commented:
@gregw I cannot see that failure in the CI results. But that would indicate an issue with the original PR #13264 not with this one.

It is because we are registering as `fillInterested` inside the `FILLING_AND_PARSING`  state, so there is a race if `onFillable()` is called before we transition back to `IDLE` state. 

So I have added another commit to this PR to fix this.

joakime commented:
@lachlan-roberts this missed the 12.0.23 release.
It isn't a a blocker issue.  
No tests on our side are failing, nor are the cometd / spring-framework / spring-boot tests of the existing 12.0.23 as well.
It's going in 12.0.24.

lachlan-roberts commented:
@joakime the issue is a race condition, so the tests just got lucky and didn't trigger it.

I really don't think we should release 12.0.23 without this fix.

joakime commented:
> @joakime the issue is a race condition, so the tests just got lucky and didn't trigger it.
> 
> I really don't think we should release 12.0.23 without this fix.

There is no test case for this supposed race condition.
There is no reproduction case for this supposed race condition.
This PR is insufficient to sneak into 12.0.23 release (which is already 6 days late)

joakime commented:
I've been able to confirm that the mentioned stacktrace exists in the `jetty-12.0.x` HEAD builds right now.

```
2025-06-30 16:37:45.140:WARN :oejuts.AdaptiveExecutionStrategy:qtp83786308-661: Task run failed
java.lang.IllegalStateException: FILLING_AND_PARSING
	at org.eclipse.jetty.websocket.core.WebSocketConnection.fillAndParse(WebSocketConnection.java:489)
	at org.eclipse.jetty.websocket.core.WebSocketConnection.onFillable(WebSocketConnection.java:381)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:322)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:99)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.runTask(AdaptiveExecutionStrategy.java:480)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.consumeTask(AdaptiveExecutionStrategy.java:443)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.tryProduce(AdaptiveExecutionStrategy.java:293)
	at org.eclipse.jetty.util.thread.strategy.AdaptiveExecutionStrategy.run(AdaptiveExecutionStrategy.java:201)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:311)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:981)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1211)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1166)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

```
2025-06-30 16:37:45.206:WARN :oejut.QueuedThreadPool:WebSocket@1004570941-2013: Job failed
java.lang.IllegalStateException: FILLING_AND_PARSING
	at org.eclipse.jetty.websocket.core.WebSocketConnection.fillAndParse(WebSocketConnection.java:489)
	at org.eclipse.jetty.websocket.core.WebSocketConnection.onFillable(WebSocketConnection.java:381)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:322)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:99)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:981)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1211)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1166)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

lachlan-roberts committed:
PR #13299 - changes from review

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

joakime commented:
Stack is still present in JDK17 build.

https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/PR-13299/4/pipeline/12

```
2025-07-01 16:03:39.292:WARN :oejut.QueuedThreadPool:WebSocket@1389194219-442: Job failed
java.lang.IllegalStateException: FILLING_AND_PARSING
	at org.eclipse.jetty.websocket.core.WebSocketConnection.fillAndParse(WebSocketConnection.java:477)
	at org.eclipse.jetty.websocket.core.WebSocketConnection.run(WebSocketConnection.java:380)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:981)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.doRunJob(QueuedThreadPool.java:1211)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1166)
	at java.base/java.lang.Thread.run(Thread.java:840)
```

gregw commented on jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/WebSocketConnection.java:
```diff
@@ -331,26 +331,20 @@ public void failed(Throwable cause)
-            networkBuffer.release();
-            networkBuffer = newNetworkBuffer(getInputBufferSize());
-        }
+        networkBuffer.release();
+        networkBuffer = newNetworkBuffer(getInputBufferSize());
```
@lachlan-roberts not for this PR, but we should do the same as HTTP/1 and reuse the existing buffer, even if retained, if there is sufficient space in it to do something useful. See `org.eclipse.jetty.server.internal.HttpConnection#parseAndFillForContent`

gregw reviewed (approved):
Looks good. But see comment about an enhancement for another PR

sbordet reviewed (changes requested):
Race condition still present when `demand()` is called from application code, and the parser thread has not hit the `finally` block in `filllAndParse()`.

gregw reviewed (changes requested):
@lachlan-roberts, Simone briefed me on the remaining race condition.  I will look today.  Ping me if you are online at all

gregw committed:
Added the extra MORE_FILLING_AND_PARSING state

This resolves the race between one thread asking for more f&p and another exiting the state.

Will investigate an alternative using ICB now.

gregw committed:
Added the extra MORE_FILLING_AND_PARSING state

This resolves the race between one thread asking for more f&p and another exiting the state.

Will investigate an alternative using ICB now.

gregw commented:
@lachlan-roberts @sbordet I tried converting the fillAndParse into an IteratingCallback.  It was surprisingly mostly successful, and most of the tests pass.   But then a few tests fail in strange ways (missing bytes from the middle of message etc), so it is not as simple as it looks.  I think replacing executes with iterates is a big change in the threading model and we should not consider it for this release cycle.  Instead I think the fix as it is now should go to the release and @lachlan-roberts can look at making it an ICB after.

lachlan-roberts committed:
PR #13299 - fixes to MORE_FILLING_AND_PARSING state

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

lachlan-roberts committed:
PR #13299 - fixes to MORE_FILLING_AND_PARSING state

Signed-off-by: Lachlan Roberts <lachlan.p.roberts@gmail.com>

joakime commented:
If the build is green, and you are happy with these changes, do merge this. 
If you cannot, say so (and why you cannot).

joakime commented:
@gregw I think Lachlan will need and approval to merge
